### DISCUSSION TITLE
cn/2024/03/22/personal-website-tutorial/

### DISCUSSION THREAD

**Discussion by giscus** [hongtaoh/hongtaoh.github.io#158] [Announcements]
# cn/2024/03/22/personal-website-tutorial/

https://hongtaoh.com/cn/2024/03/22/personal-website-tutorial/

**Comment by Finderaha** [comment]
在网站主题那一步，我遇到一个问题：在hugo server -D后，系统提示如下：
Watching for changes in D:\个人博客\quickstart\{archetypes,assets,content,data,i18n,layouts,static}
Watching for config changes in D:\个人博客\quickstart\hugo.toml
Start building sites …
hugo v0.124.1-db083b05f16c945fec04f745f0ca8640560cf1ec windows/amd64 BuildDate=2024-03-20T11:40:10Z VendorInfo=gohugoio

WARN  found no layout file for "html" for kind "section": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
WARN  found no layout file for "html" for kind "term": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
WARN  found no layout file for "html" for kind "taxonomy": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
WARN  found no layout file for "html" for kind "page": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
WARN  found no layout file for "html" for kind "home": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.

                   | EN
-------------------+-----
  Pages            | 16
  Paginator pages  |  0
  Non-page files   |  0
  Static files     |  0
  Processed images |  0
  Aliases          |  0
  Cleaned          |  0

Built in 28 ms
Environment: "development"
Serving pages from disk
Running in Fast Render Mode. For full rebuilds on change: hugo server --disableFastRender
Web Server is available at http://localhost:1313/ (bind address 127.0.0.1)
Press Ctrl+C to stop
访问网页时，提示Page Not Found，提示信息我理解是：Hugo 在寻找 "home"、"section"、"page"、"term" 和 "taxonomy" 这些内容类型的布局（layout）文件时未能找到对应的 .html 模板文件。不知道为啥会出现这个问题，同时也不知道具体应该怎么解决？求解惑。

**Reply by Finderaha** [reply]
前面我都是严格按照部门执行的，感觉上应该是没问题的。

**Reply by hongtaoh** [reply]
这肯定又是操作系统不同导致的。可是我手头没有 windows 系统的电脑，暂时没办法帮你排查原因。

**Reply by hongtaoh** [reply]
我想了一下，应该是这两行出了问题：

```bash
cp -r themes/hugo-ht/exampleSite/* .
cp -r themes/hugo-ht/archetypes . 
```

刚才问了一下 chatgpt，它给我的建议是，在 Windows 系统中，如果你用的是 git bash， 那上面的代码是可以的。如果你用的是 Powershell, 那么改成

```bash
Copy-Item -Recurse themes\hugo-ht\exampleSite\* .
Copy-Item -Recurse themes\hugo-ht\archetypes .
```

如果你用的是 Command Prompt，那两行改成

```bash
xcopy /E /I themes\hugo-ht\exampleSite\* .
xcopy /E /I themes\hugo-ht\archetypes .
```

你看看是不是有效果。

**Reply by Finderaha** [reply]
在 windows 上我用的是 git bash，后面我又在 mac 上尝试了一下，到这里时遇到了同样的问题。也是提示：Page Not Found

**Reply by hongtaoh** [reply]
现在应该可以了！加了一行：

```
rm hugo.toml
```

因为这个和 `config.toml` 冲突了。

**Reply by hongtaoh** [reply]
怎么样，问题解决了吗？

**Reply by Finderaha** [reply]
嗯，解决了，可以正常显示了。我不理解你说的冲突是什么意思？

**Reply by hongtaoh** [reply]
好问题。Hugo 目前默认的配置文件是 `hugo.toml`，但是我做的这个主题用的还是之前 Hugo 默认的 `config.toml`。我本来想改成 `hugo.toml` 但是 Vercel 默认的是 `config.toml`，这么一来对新手又多了一道障碍，所以我没有改。

在没有加 `rm hugo.toml` 之前，hugo 文件夹里同时存在两者，`hugo server -D` 的时候，hugo 用的是默认的空荡荡的 `hugo.toml`，而不是我已经配置好的 `config.toml`，故而报错。

**Comment by Finderaha** [comment]
在执行上述流程中遇到的几个问题的说明：
1. git submodule add https://github.com/hongtaoh/hugo-ht themes/hugo-ht
    这里我遇到了 2 个问题，
     * 因为我之前配置了ssh，在复制 git 项目时使用了ssh链接，这里应该使用https 链接。
     * 发生报错：fatal: 'themes/hugo-ht' already exists in the index 你正在尝试添加一个已经跟踪的目录或文件到暂存区（index）。这里的解决方法是：
       删除缓存区内容，sudo rm -r .git，需要输入一下电脑密码。删除后需要重新开始 git init
2. 在 vercel 中Deploy时，总是提示构建错误，sh: line 1: hugo: command not found 具体原因我不太清楚，解决方案是：在 vercel 的项目 setting 页面-general-Build & Development Settings中，build command 选项内容改为：npm i -g hugo即可。

**Reply by Finderaha** [reply]
更新一下问题 2，上述解决方案会造成后续更新无法生肖。应该使用的解决方案是，在项目 setting - Environment Variables中，设置 hugo 的版本信息，在 key 中写：HUGO_VERSION Value中填写电脑上的hugo version（方法是打开终端 hugo version 即可看到）。
出现这个问题的原因是 vercel 上默认的 hugo version 比较老，导致在构建时总是报错，解决方案就是上面提到的在 vercel 中声明一下 hugo version即可。

**Reply by hongtaoh** [reply]
奇怪的是为什么我这里不会碰到类似的问题。

**Reply by Finderaha** [reply]
可能你的电脑 hugo 版本也不高。

**Comment by Finderaha** [comment]
提 2 个文档之外的问题：
1. 我想以中文为主，所以按照你 21 年的教程做了配置，配之后确实默认导航变成了 主页、日志、关于、English，但是有一个问题，当点击“主页”时，导航会自动切到英文导航（about、posts、中文），如何能保持在中文导航。
2. 我想修改一下网站左上角的 logo 成自己的，尝试的方法是修改本地 media 中的 logo 为自己想要的，同时保持名称和原来的一致，然后推送到 git 上，但是无法成功。想知道正确的方法是什么？

这是我网站的地址：https://myblog-six-pink.vercel.app/

**Reply by hongtaoh** [reply]
好问题。

1. 我自己按照 [21 年教程的相关部分](https://hongtaoh.com/cn/2021/03/02/personal-website-tutorial/#%E4%B8%AD%E6%96%87%E4%B8%BA%E4%B8%BB%E8%8B%B1%E6%96%87%E4%B8%BA%E8%BE%85)设置了之后，没有出现你说的这个问题。
2. 你可以试试修改 `themes/hugo-ht/static/media` 中的 `hugo-logo.png`。（这个是 `config.toml` 中的 favicon = "media/hugo-logo.png"）

**Reply by Finderaha** [reply]
更新一下 2 个问题：
1. 这个问题如果在本地 hugo server -D不会出现，git 上看也是新的，但 vercel 有问题。原因是：vercel 认为 themes/hugo-ht 是我引用的他人项目，所以没有拉取新的内容导致。
解决方法是：我复制了一个 hugo-ht，改名叫 hugo-zj，然后修改了一下 config中theme，就好了。
我听他们说也可以把 hugo-ht 中隐藏的 .git删除掉，vercel 不认为这是一个独立的他人项目，但是我没有试。

2. 第 2 个问题是我偷懒，直接把自己想用的 logo 改成原来 logo 的名称 hugo-logo，导致的。原因是，vercel 在检查资源的时候，没有检查文件，只是检查了文件名，因为时同名，所以认为没有更新，使用了旧的文件导致的。
解决方法：改了个新的文件名，然后改动了一下 config 和 theme/hugo-zj/layouts/partials/nav.html文件中的logo 指向就好了。

解决方法肯定是有效的，但是不确定这 2 个问题的原因是否真是如此。

**Comment by Finderaha** [comment]
我看到你的标签能显示在文章下面，但是我写的文章没有办法显示标签。求解。

**Reply by hongtaoh** [reply]
现在可以了。我把 hugo-ht 更新了，你用最新的试试。改动主要是在 `layouts/_default/single.html`。一篇文章可以只有单个 tag:

```html
tags: prose
```

也可以有多个：

```html
tags:
  - prose
  - article
```

**Reply by hongtaoh** [reply]
感觉你解决了很多问题。不知道能否写一篇博客记录一下你整个的建站已经解决问题的过程，这样肯定会帮到很多人。

**Reply by Finderaha** [reply]
嗯，好的。等我把一些框架的问题都搞清楚了，写一下这段时间遇到的问题。

**Reply by Finderaha** [reply]
梳理了一下之前遇到的问题，记录如下：https://www.finderaha.com/cn/2024/04/14/beiying/

**Reply by hongtaoh** [reply]
你的这篇博客参数似乎没有改 slug，所以最后的网址是 beiying。你可以改成 blog，就成了 https://www.finderaha.com/cn/2024/04/14/blog 。

[这里](https://www.finderaha.com/cn/2024/04/14/beiying/#1-%E4%B8%BA%E4%BB%80%E4%B9%88%E8%A6%81%E6%89%A7%E8%A1%8C%E4%B8%8B%E9%9D%A2%E8%BF%99%E6%AE%B5%E4%BB%A3%E7%A0%81) 我比较奇怪，你似乎用的是之前教程的代码？

关于 git submodule hugo-ht 这部分带来的问题，我自己之后用一个新的 github username 来试一下，看起来问题挺多的。

**Reply by Finderaha** [reply]
嗯，slug 之前没理解是做什么的。

你说的奇怪的地方，确实是使用之前的教程代码引发的。一方面是因为我先看到你21 年的教程，后面才注意到你 24 年更新的这一篇。可以考虑在 21 年教程里 创建 quickstart后写一个新链接。另一方面你在 24 年教程里“把之前教程「网站主题」之前的部分看完”我会错意了，以为包括网站主题部分。

新的 github username，我比较顺利🤣。

**Reply by hongtaoh** [reply]
再来一个黑科技：https://hongtaoh.com/cn/2021/10/13/hugo-aliases/

这个可以解决 https://www.finderaha.com/cn/2024/04/14/beiying/ 404 的错误，可以直接重新定向到你目前的网址。

**Comment by Finderaha** [comment]
遇到 2 个问题：
1. tags 我更新了hugo-ht后，确实显示了，但是点击会提示 404
2. 评论我按照 21 年教程执行后，并没有显示。

**Reply by hongtaoh** [reply]
1. 把 `config.toml` 里的 baseURL 改成你目前的网址 https://www.finderaha.com/ 
2. 重装一下 hugo-ht，把 `vercel.json` 里的版本号改成你的，然后看一下我[更新的教程](https://hongtaoh.com/cn/2024/03/22/personal-website-tutorial/#%E6%B7%BB%E5%8A%A0%E8%AF%84%E8%AE%BA)

**Comment by CyrusYip** [comment]
我想到一个更低门槛，不需要命令行的方法。

1. 把你的实例网站代码弄成[模板仓库](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository)，用户可以一键创建网站。
2. 把网站部署到 Vercel 或者其他 Jamstack 服务
3. 用在线 CMS，连接 GitHub 仓库后就是在线 Hugo 编辑器。我觉得 forestry.io 不错，不过它现在成了 https://tina.io/ ，也不知道现在好不好用了。
4. 修改配置可以直接用 GitHub 的在线编辑器。

**Reply by hongtaoh** [reply]
嗯嗯。感觉还是有点复杂 hhh。我觉得真的有兴趣的人会去学一些代码的。另外，我觉得目前貌似国内没有一个面向大众的博客平台？

**Reply by CyrusYip** [reply]
> 嗯嗯。感觉还是有点复杂 hhh。我觉得真的有兴趣的人会去学一些代码的。

我这个方法面向不懂技术的人，普通人也有建博客的需求，但是常规方法对他们来说可能太难了，要先学习很多东西。

> 另外，我觉得目前貌似国内没有一个面向大众的博客平台？

简书、博客园？我觉得博客在国内不流行了，写长文更多人用微信公众号。

**Reply by hongtaoh** [reply]
我觉得可以！你有空试试吗？弄好了我帮忙推广。

很可惜。那么好的内容都被腾讯垄断了。

**Reply by CyrusYip** [reply]
我觉得这个教程的难点在于解答读者的问题，我没那么多精力。可能我会参考益辉的《[是莽撞人就来单挑：教学过程中与学员实时分享代码文件](https://yihui.org/cn/2020/07/livecode/)》，把这些想法写到博客，等有缘人来写吧。

**Comment by daoyuanw** [comment]
hugo serve D后，打开http://localhost:1313/ 提示localhost 拒绝了我们的连接请求，是什么原因呢

**Reply by hongtaoh** [reply]
你试一下这个？https://blog.csdn.net/weixin_46931860/article/details/124108595

**Reply by daoyuanw** [reply]
谢谢~已解决，关掉防火墙就好了。2分钟的视频，我follow下来花了一下午，过程很曲折，但是终于建好了，谢谢鸿涛。附上我的主页（虽然还没开始写blog）https://my-blog-seven-beige.vercel.app/

**Comment by DongFengto** [comment]
谢谢博主的分享！另外如果只需一个个人页面，类似很多高校老师用的那种，可以参照这篇文章https://blog.csdn.net/qd1813100174/article/details/128604858，直接fork模版，相当方便。

**Comment by yulinyang-linda** [comment]
请教博主一个问题：在vercel deploy那一步总是显示“The provided GitHub repository does not contain the requested branch or commit reference. Please ensure the repository is not empty.” 应该怎么解决呢？

**Reply by yulinyang-linda** [reply]
再补充一点，vercel没有自动将我的project识别为hugo

**Reply by hongtaoh** [reply]
我看了一下，你的 github repository 都是空的。应该是你哪里做错了导致没有上传到 github。具体哪里错了我不清楚。

**Reply by yulinyang-linda** [reply]
感谢查看！我又检查了一下 代码应该是没有问题，都是按照博主的步骤来的。我发现我在新建仓库那一系列代码运行结束之后，终端返回的最后一行代码是"Username for 'https://github.com':"，会不会是这里出现了问题呢？另外，我的文本编辑器没有安装成功，但看代码中似乎没有需要用到它的地方，不知道这是否会有影响？

**Reply by hongtaoh** [reply]
文本编辑器那个不重要，你用 vs code 也可以。

>"Username for https://github.com':

这里是问题的关键，你需要输入你的 github 账号用户名和密码。但是因为 github [登录方式的变更](https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/)，你用密码大概率不行，而是需要用 personal access token，具体怎么操作你可以[看这里](https://hongtaoh.com/en/2021/08/03/github-authentication/)。

**Reply by yulinyang-linda** [reply]
问题解决啦，不过这个personal access token没有派上用场，最后是下载了GCM解决的。我的网站也基本搭好了，非常感谢鸿涛师兄！另外我还在我的[日志](https://yulinyang.vercel.app/cn/2024/07/20/%E8%AE%B0%E4%B8%AA%E4%BA%BA%E7%BD%91%E7%AB%99%E7%9A%84%E5%BB%BA%E7%AB%8B/)里增加了一些问题和解决方式，希望可以帮助到后来者。

**Comment by ghost** [comment]
学长您好，我发现用vercel建成网站之后，您在原帖中用namecheap换域名的教程可能需要更改，我研究了一下，如果新版教程需要更换域名的话，首先需要登入vercel中的domain，add新增的购买的域名，之后vercel会报错，并提示需要先在namecheap中配置dns，并列出所需要变更的A-record：76.76.21.21，（这串数字我的vercel跳出来的不知道，别人一不一样），ttl我选择了Automatic。其实我跟这老教程也添加了workflows/gh-pages.yml文件，不知道这是不是必须的哈哈

**Reply by hongtaoh** [reply]
感谢分享。原来的教程是根据 github pages 做的，和 vercel 不同。

**Reply by hongtaoh** [reply]
你的网站我打不开。

**Comment by daoyuanw** [comment]
我又来了，摆脱单机的意思是，终于有人访问我的blog了，不是互联网孤岛了。看了下代码果然没有改启用giscuss的script，但是本地改完后git bash了后我那篇帖子打不开了，提示500: INTERNAL_SERVER_ERROR。

**Reply by hongtaoh** [reply]
帮你改好了：https://github.com/daoyuanw/my-blog/pull/3

**Comment by daoyuanw** [comment]
我发现修改已发布的帖子内容时，最后一次修改时间仍然是初始创建文件时间。请问这个时间可以在哪里修改呢？

**Reply by hongtaoh** [reply]
[这里](https://github.com/daoyuanw/my-blog/blob/main/themes/hugo-ht/layouts/_default/single.html#L49) 需要修改。改成 [这个](https://github.com/hongtaoh/hongtaoh.github.io/blob/sources/themes/hugo-ht/layouts/_default/single.html#L42)。

**Comment by chuxinyuan** [comment]
Vercel 有段时间国内访问不了，被 Netlify 和 Vercel 撵来撵去的，后来买了服务器，再后来我才知道其实 Cloudflare 也是很棒的选择。但是折腾服务器也学到了很多知识。有一段时间沉迷于照着益辉的博客改造自己的博客，加上各种酷炫功能，再然后做各种优化提高访问速度。最近才开始不再折腾，做点内容分享。

**Reply by hongtaoh** [reply]
嗯嗯。我现在就用的 cloudflare。这个在国内比较快。

>最近才开始不再折腾，做点内容分享。

这很正常。
